### ISSUE TITLE
Fetching from Local Data Store  using PFQuery with "selectKeys" not working on iOS 9

### ISSUE THREAD

**Issue by asmusto** [ParsePlatform/Parse-SDK-iOS-OSX#386]
Hello guys,

As the title mentioned, after pinning to local data store, objects that are queried using "selectKeys", the objects can not be fetched on iOS 9

The Steps are like this 
1. Create PFQuery like this :
PFQuery * q = [PFQuery queryWithClassName:@"someClass"];
[q selectKeys:@[@"col1",@"col2"]];
2. Fetch from local datastore. If not exist, download from parse and then pin to local datastore
3. The next time it is fetched from local datastore, it supposed to exist.

In iOS 9 both simulator and real device, this problem occurs. But in iOS 8 simulator, the objects are successfully saved and fetched.
However, when I removed the "selectKeys", and retrieved all columns, it works fine both in iOS 9 simulator and device.

**Comment by richardjrossiii** [comment]
Thanks for the report, this sounds super strange. I'll look into it.

**Comment by FreudGit** [comment]
Related to this?
https://developers.facebook.com/bugs/560039110805597/

**Comment by richardjrossiii** [comment]
@FreudGit I don't believe so, as that bug report doesn't mention iOS version at all.

@asmusto I just ran the following code sniper locally, and everything seems to be working on iOS 9, simulator and device... Can we get a more concrete code snippet?

```objc

#import <Parse/Parse.h>

@interface TestObject : PFObject<PFSubclassing>

@property NSString *foo;
@property NSString *bar;

@end

@implementation TestObject

@dynamic foo, bar;

+ (NSString *)parseClassName {
    return @"TestObj";
}

@end

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        [TestObject registerSubclass];

        for (int i = 0; i < 15; i++) {
            TestObject *obj = [TestObject new];
            obj.foo = [NSString stringWithFormat:@"Foo %i", i];
            obj.bar = [NSString stringWithFormat:@"Bar %i", i];

            [obj pin];
        }

        PFQuery *query = [TestObject query];
        [query selectKeys:@[ @"foo", @"bar" ]];
        [query fromLocalDatastore];

        NSArray *objects = [query findObjects];
        NSLog(@"%@", objects);

        [TestObject unpinAll:objects];
    });
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end
```

**Comment by asmusto** [comment]
Hi @richardjrossiii 

Thank you for prompt reply :)

Your code is correct, but that's only for data stored locally and the columns are all retrieved.
I've changed your code to something like this :

```objc
#import "TestObject.h"

@implementation TestObject

@dynamic foo, bar;

+ (NSString *)parseClassName {
    return @"TestObj";
}

+ (void) uploadAndPinToLocal
{
    [TestObject registerSubclass];
    
    [PFObject unpinAllObjects];

    dispatch_async(dispatch_get_main_queue(), ^{
        
        NSMutableArray * objects = [NSMutableArray new];
        for (int i = 0; i < 5; i++) {
            TestObject *obj = [TestObject new];
            obj.foo = [NSString stringWithFormat:@"Foo %i", i];
            obj.bar = [NSString stringWithFormat:@"Bar %i", i];
            
            [objects addObject:obj];
        }
        
        [PFObject saveAllInBackground:objects block:^(BOOL succeeded, NSError * _Nullable error) {
            
            PFQuery *query = [TestObject query];
            [query selectKeys:@[ @"foo"]]; // remove this code or select All keys and it will work
            [query findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
                
                if (objects) {
                    [PFObject pinAllInBackground:objects block:^(BOOL succeeded, NSError * _Nullable error) {
                        
                        NSLog(@"success pin to background %d", succeeded);

                    }];
                }
                
            }];

        }];
        
        
    });
}

+ (void) getFromLocal
{
    [TestObject registerSubclass];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        PFQuery *query = [TestObject query];
        [query selectKeys:@[ @"foo"]]; // with or without selectKeys doesn't matter
        [query fromLocalDatastore];
        
        NSArray * objects = [query findObjects];
        NSLog(@"get from local %@", objects);

    });
}

```

And in ViewController.m, try this to upload the objects and pin it to local datastore, and then after successfully finished, quit the app

```objc
- (void)viewDidLoad {
    [super viewDidLoad];
    [TestObject uploadAndPinToLocal];   
}
```

After quitting the app, change the code to this (to get the data from pinned local datastore). And then run the app again.
```objc
- (void)viewDidLoad {
    [super viewDidLoad];
    [TestObject getFromLocal];   
}
```

Note that the problem won't happen if PFQuery all keys are added when fetching from server and then pin locally.

Thanks! :)

**Comment by FreudGit** [comment]
@richardjrossiii Please start from the end of this page:  https://developers.facebook.com/bugs/560039110805597/
"They plan to address it long term. For now, please do not use  [query selectKeys: for saving to local datastore. 
We are going to update the docs with this limitation. I will keep this bug open to track the changes in the doc. "

"Tushar Arora ·  · Facebook Team
you cannot select only 2 columns. you will have to pull all columns . -  that's the limitation."

Long time ago, I find it very weird, but Tushar close the report.  Even not possible to get the doc updated.

**Comment by asmusto** [comment]
Hi @FreudGit 
Thanks for the info. I re-checked with iOS 8 simulator. I said earlier that this doesn't happen in iOS 8, but I tested again and it did happen. So I guess it is really the limitation at the moment and to not use selected keys for saving to local data store.
Maybe it's related with saving the "non-existing" data for the columns that are not selected.
I vaguely remember a while ago, trying to use the "non-existing-data" column and the app just freeze.
Ah well, at least now I know how to fix this for the moment.
Thanks again all :)

**Comment by FreudGit** [comment]
@richardjrossiii I still suggest to update the doc OR to solve this.  ;)

**Comment by sirnacnud** [comment]
Did this limitation ever get fixed? I'm running in to it with version 1.11.0 of the SDK.

**Comment by ProximalPod** [comment]
This is still an issue. However if you move your query to cloud code and use query.select() there, you can successfully pin the result and retrieve it from the local datastore.
